frontend/ui/widget/button.lua

```diff
@@ -123,7 +123,9 @@ end
 function Button:setText(text, width)
     self.text = text
     self.width = width
+    local dimem = self[1].dimen
     self:init()
+    self[1].dimen = dimem
```

Comment:
Is that with no risk?
Some button with small text, when given some larger text, would not increase its width then?
Like in frontend/ui/widget/menu.lua: `self.page_info_text:setText(util.template(_("page %1 of %2"), self.page, self.page_num))` could become truncated/misaligned when you change from `page 1 of 33` to `page 24 of 33` ?
(not tested, just thinking.)
Might be worth making it an option if it's really needed.
Or, as there is already a width parameter, may be using it like that would work?
```diff
- self.page_cnt[1]:setText(curr .. "/" .. total)
+ self.page_cnt[1]:setText(curr .. "/" .. total, self.page_cnt[1].width) -- or self.page_cnt[1].dimen.w)
```
Or do what you do if you are provide a `width=true` (instead of a size), true meaning: keep the existing dimen.

Reply:
I need it because without this line:

`self[1].dimen = dimem
`

I get error (crash) after tap on `▼`

```
./luajit: frontend/ui/widget/button.lua:228: attempt to index field 'dimen' (a nil value)
stack traceback:
	frontend/ui/widget/button.lua:228: in function 'action'
	frontend/ui/uimanager.lua:750: in function '_checkTasks'
	frontend/ui/uimanager.lua:1033: in function 'handleInput'
	frontend/ui/uimanager.lua:1119: in function 'run'
	./reader.lua:260: in main chunk
	[C]: at 0x563cba790964

```
self[1].dimen is null but I really don't know why :(
Any idea how to do it correctly?

Reply:
button.lua:228 is code related to flashing the button when tap: you tap, its inverted to black, and a scheduled action is set to put it back to white. But in the meantime, with setText, you have re-initing the button, so, when in the scheduled action some milliseconds later, `self[1] = self.frame` is another object, which might not yet have a dimen (or it's the old object that has already been partly destroyed).
So, it might then better be fixed in the scheduled callback: it's its job to check that what it has to invert back still exists.
So, may be in there:
https://github.com/koreader/koreader/blob/3d696d9d87c9a068667c956db38f97ddcc26e0c9/frontend/ui/widget/button.lua#L225-L228
add
```lua
 UIManager:tickAfterNext(function() 
+    if not self[1] or not self[1].invert or not self[1].dimen then
+        -- widget no more there (destroyed, re-init'ed by setText(), or not inverted: nothing to invert back
+        return
+    end
     self.callback()
     [...]
```

Reply:
Thanks.
It works perfectly but I have to move self.callback() on top.
`self.callback()` "breaks" `self[1].dimen`

```lua
            UIManager:tickAfterNext(function()              
                self.callback()
                if not self[1] or not self[1].invert or not self[1].dimen then
                    -- widget no more there (destroyed, re-init'ed by setText(), or not inverted: nothing to invert back
                    return
                end
                self[1].invert = false
                UIManager:widgetRepaint(self[1], self[1].dimen.x, self[1].dimen.y)
                [...]

```